src/wp-includes/fonts/class-wp-font-family-utils.php

```diff
@@ -0,0 +1,89 @@
+<?php
+/**
+ * Fonts Family Utils class.
+ *
+ * This file contains utils fot Font Family class.
+ *
+ * @package    WordPress
+ * @subpackage Font Library
+ * @since      6.4.0
+ */
+
+/**
+ * A class of utilities for working with the Font Library.
+ *
+ * @since 6.4.0
+ */
+class WP_Font_Family_Utils {
+
+	/**
+	 * Generates a filename for a font face asset.
+	 *
+	 * Creates a filename for a font face asset using font family, style, weight and
+	 * extension information.
+	 *
+	 * @since 6.4.0
+	 *
+	 * @param string $font_slug The font slug to use in the filename.
+	 * @param array  $font_face The font face array containing 'fontFamily', 'fontStyle', and
+	 *                          'fontWeight' attributes.
+	 * @param string $url       The URL of the font face asset, used to derive the file extension.
+	 * @param string $suffix    Optional. The suffix added to the resulting filename. Default empty string.
+	 * @return string The generated filename for the font face asset.
+	 */
+	public static function get_filename_from_font_face( $font_slug, $font_face, $url, $suffix = '' ) {
+		$extension = pathinfo( $url, PATHINFO_EXTENSION );
+		$filename  = "{$font_slug}_{$font_face['fontStyle']}_{$font_face['fontWeight']}";
+		if ( '' !== $suffix ) {
+			$filename .= "_{$suffix}";
+		}
+
+		return sanitize_file_name( "{$filename}.{$extension}" );
+	}
+
+	/**
+	 * Merges two fonts and their font faces.
+	 *
+	 * @since 6.4.0
+	 *
+	 * @param array $font1 The first font to merge.
+	 * @param array $font2 The second font to merge.
+	 * @return array|WP_Error The merged font or WP_Error if the fonts have different slugs.
+	 */
+	public static function merge_fonts_data( $font1, $font2 ) {
+		if ( $font1['slug'] !== $font2['slug'] ) {
+			return new WP_Error(
+				'fonts_must_have_same_slug',
+				__( 'Fonts must have the same slug to be merged.' )
+			);
+		}
+
+		$font_faces_1      = isset( $font1['fontFace'] ) ? $font1['fontFace'] : array();
+		$font_faces_2      = isset( $font2['fontFace'] ) ? $font2['fontFace'] : array();
+		$merged_font_faces = array_merge( $font_faces_1, $font_faces_2 );
+
+		$serialized_faces        = array_map( 'serialize', $merged_font_faces );
+		$unique_serialized_faces = array_unique( $serialized_faces );
+		$unique_faces            = array_map( 'unserialize', $unique_serialized_faces );
```

Comment:
Seralization can be expensive compute wise. Was json_decode considered?

Reply:
Also, this looks like a very JS-esque kinda way to make an array unique. Nested foreach loops or something would be more performant. Or if fonts are actual class instances instead of arrays, `array_unique()` + `__toString()` could be used.

Reply:
If this is JSON data, just serializing like this is incorrect since it doesn't follow JSON semantics regarding key order. We have `rest_stabilize_value` in Core for that.

Is this function called in a hot-path? I.e. does it happen on most page loads? Or is this something called during the ingestion(?) of a font, i.e. once. If it is in a hot-path I agree this should be done more performantly. If this only happens during the initial download, then I think serializing is fine, that's how we validate uniqueness in the REST API, see `rest_validate_array_contains_unique_items`.

Reply:
> s this function called in a hot-path? I.e. does it happen on most page loads? Or is this something called during the ingestion(?) of a font, i.e. once.

This is called only when a new font is installed and NOT on every page load.

Reply:
I see no references to `serialize` in the current Gutenberg `trunk` files for the Font Library feature.

https://github.com/WordPress/gutenberg/tree/trunk/lib/compat/wordpress-6.5/fonts

Some "merging" happens [here](https://github.com/WordPress/gutenberg/blob/066f8777b113ca8a71c79deb052d1deb23fe9860/lib/compat/wordpress-6.5/fonts/class-wp-rest-font-families-controller.php#L509-L519) but the code has evolved significantly.